### ISSUE TITLE
`torch.optim.lr_scheduler.LRScheduler`: lack of documentation

### ISSUE THREAD

**Issue by kuraga** [pytorch/pytorch#120735]
### 📚 The doc issue

The `torch.optim.lr_scheduler.LRScheduler` class [doesn't have](https://github.com/pytorch/pytorch/blob/v2.2.1/torch/optim/lr_scheduler.py#L33) docstring.

Some of its methods don't.

Because of this we don't know:

* What the constructor's `last_epoch` argument is.
* Which arguments has the `.step()` method.
* That `.get_last_lr()` and `.get_lr()` are totally different methods despite of naming.

In contrast, `torch.optim.Optimizer` has.

**Comment by janeyx99** [comment]
We would accept a PR with a fix so that these do show up in our docs!

**Comment by kuraga** [comment]
Oh, no, sorry, I understand that too much.
I even didn't know scheduler does `.step()` at initialization...

That's (one of) why IMHO it's an issue :)

**Comment by kuraga** [comment]
On p.1:

> What the constructor's `last_epoch` argument is.
> And does epoch start from `0` or `1`?
> Also there are two terms - "epoch", "step" - are they the same?

There is documentation line almost in each `torch.optim.lr_scheduler.LRScheduler` subclass:

> ```
> # When last_epoch=-1, sets initial lr as lr.
> ```

Also was discussed it here: https://discuss.pytorch.org/t/when-last-epoch-1-sets-initial-lr-as-lr/197812
According to the discussion:

1. The sense of this line is not totally understandable.
2. The `last_epoch` argument just ["scrolls toward"](https://github.com/pytorch/pytorch/blob/v2.2.1/torch/optim/lr_scheduler.py#L52-L52) `last_epoch+1` steps.
3. This phrase belongs to the `torch.optim.lr_scheduler.LRScheduler` class itself instead of every subclass (where it is written at).
4. There is no super-special case on `last_epoch=-1`.

So this phrase should be rewritten.

**Comment by kuraga** [comment]
On p. 3:

> `.get_last_lr()` and `.get_lr()` are totally different methods despite of naming.
> There are also `._get_closed_form_lr()` methods, hm...

There are two different behaviors:

https://github.com/pytorch/pytorch/blob/v2.2.1/torch/optim/lr_scheduler.py#L158-L161

In subclasses, `._get_closed_form_lr()`s and `.get_lr()`s are algorithmically different.
(As an example, the first relies on `.base_lrs` information (see also p.2); the latter doesn't.)

This difference isn't documented.
(And maybe not tested whether they provide the same values.)

**Comment by kuraga** [comment]
On pp. 4 and 6:

> The constructor [does `.step()`](https://github.com/pytorch/pytorch/blob/v2.2.1/torch/optim/lr_scheduler.py#L85-L91) itself via `_initial_step()` method.
> What does `.step()` do? Does it modify (in some way) the `optimizer`? (See also p.2.)

So after the scheduler object, the "first" (i.e. `<last_epoch argument>+``-th) step had been done.
Also the `optimizer`'s `'lr'`s had been [replaced](https://github.com/pytorch/pytorch/blob/v2.2.1/torch/optim/lr_scheduler.py#L163-L165)).
None of this behavior is documented.

A "funny" consequence are there in the `torch.optim.lr_scheduler.SequentialLR` class.
It [remakes](https://github.com/pytorch/pytorch/blob/v2.2.1/torch/optim/lr_scheduler.py#L693-L704)
these steps in its own way (i.e. "`reset learning rates back to initial values"`).
The code has some correct comments about this проворачивание фарша назад but publically the `SequentialLR` behavior is not documented.

**Comment by janeyx99** [comment]
https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.LRScheduler.html should be revamped to address these points.
